xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java

```diff
@@ -432,6 +448,79 @@ int pick() {
     }
   }
 
+  /*
+   * Implementation of Static Stride Scheduler
+   * Replaces EDFScheduler
+   * 
+   * go/static-stride-scheduler
+   */
+  @VisibleForTesting
+  static final class StaticStrideScheduler {
+    private final int[] scaledWeights;
+    private final int sizeDivisor;
+    private final AtomicInteger sequence;
+    private static final int K_MAX_WEIGHT = 65535; // uint16? can be uint8
+    private static final int UINT32_MAX = 429967295; // max value for uint32
+
+    StaticStrideScheduler(float[] weights) {
+      int numChannels = weights.length;
+      int numZeroWeightChannels = 0;
+      double sumWeight = 0;
+      float maxWeight = 0;
+      for (float weight : weights) {
+        if (weight < 0.0001) { // just equal to 0?
+          numZeroWeightChannels++;
+        } else {
+          sumWeight += weight;
+          maxWeight = Math.max(weight, maxWeight);  
+        }
+      }
+
+      checkArgument(numChannels >= 1, "Couldn't build scheduler: requires at least one weight");
+
+      double scalingFactor = K_MAX_WEIGHT / maxWeight;
+      int meanWeight = numZeroWeightChannels == numChannels ? 1 :
+          (int) Math.round(scalingFactor * sumWeight / (numChannels - numZeroWeightChannels));
+
+      // scales weights s.t. max(weights) == K_MAX_WEIGHT, meanWeight is scaled accordingly
+      int[] scaledWeights = new int[numChannels];
+      for (int i = 0; i < numChannels; i++) {
+        if (weights[i] < 0.0001) { // just equal to 0?
+          scaledWeights[i] = meanWeight;
+        } else {
+          scaledWeights[i] = (int) Math.round(weights[i] * scalingFactor);
+        }
+      }
+
+      this.scaledWeights = scaledWeights;
+      this.sizeDivisor = numChannels;
+      // this.sequence = new AtomicInteger((int) (Math.random() * UINT32_MAX));
+      this.sequence = new AtomicInteger(0);
+      // optimization: isn't sequence guaranteed to be in the first generation?
+      // failing test cases when initialized to non-zero value
+    }
+
+    private int nextSequence() {
+      return this.sequence.getAndUpdate(seq -> ((seq + 1) % UINT32_MAX));
```

Comment:
After the seq reaches Integer.MAX_VALUE , it rolls over and this becomes -1. The seq should be non negative.

The staticScheduler lives in the picker, and there might be multiple instances of picker in the system, so ideally the seq should be global among all the pickers. Alternatively, we need to randomize the weight during the static scheduler construction time.

Reply:
Should be fixed.

Reply:
Since you are storing 32 bits, instead of AtomicLong, use AtomicInteger. Then, you just need to cast _up_ to `long` as you read it out. That can be done with [`Integer.toUnsignedLong()`](https://docs.oracle.com/javase/8/docs/api/java/lang/Integer.html#toUnsignedLong-int-) or `seq & UINT32_MAX`. Masking instead of modulus makes it easier to avoid negative results. Then, instead of getAndUpdate(), just use getAndIncrement(). So this becomes:

```java
private final AtomicInteger sequence;

private long nextSequence() {
    return Integer.toUnsignedLong(sequence.getAndIncrement());
}
```

Note that I am paying attention to & UINT32_MAX being different from % UINT32_MAX. But we don't care about the difference here.

Reply:
I was under the impression that AtomicInteger was signed, which would mean that it cannot support the max value for an unsigned 32 bit integer. Is this not the case?

Reply:
It still has 32 bits and addition (the increment) is the same for signed and unsigned numbers. So the bits are what we'd hope they would be, but Java just won't treat them right in math. When casting to a larger integer size, then only difference between signed and unsigned is whether you sign extend: whether you fill the new bits on the left with 0s or 1s. For unsigned, you always use 0s. For signed, you copy the sign bit to the upper bits. So if you want unsigned conversion but only have signed, you do a signed conversion and then force the top bits to be 0s with the bitwise AND.

We could use AtomicLong and mask out the lower bits, but if we are masking the bits then AtomicInteger is just as good.